### ISSUE TITLE
Add `selector-no-invalid`

### ISSUE THREAD

**Issue by alisonailea** [stylelint/stylelint#3480]
## The Problem
Developers on my team occasionally add extra parenthesis to the end of custom-properties, media-queries, and urls which are not caught by other linting rules and which can cause errors when that CSS is used in other projects.

## The Solution
A new rule that checks for extra parenthesis at the end of media queries and custom-properties

### Notes
I have a branch in my local clone of Stylelint with this rule already partially written if it is something the community wants I'm happy to finish and submit the PR

### Tape test (based off no-extra-semicolons)
```js
"use strict";

const rule = require("..");
const { messages, ruleName } = rule;

testRule(rule, {
  ruleName,
  config: [true],

  accept: [
    {
      code: "a { background: url() }"
    },
    {
      code: "a { background: url('') }"
    },
    {
      code: "a { background: url( ) }"
    },
    {
      code: "a { color: var(--prop) }"
    },
    {
      code: "a { color: (); }"
    },
    {
      code: "a { color: var(--prop);}"
    },
    {
      code: "@media (min-width: var(--prop)) {a { color: var(--prop); }}"
    },
    {
      code: "/* a { color: var(--prop);} */"
    },
    {
      code: "/* a { color: var(--prop)));}; */"
    },
    {
      code: "/* ()comment)) words))(( */"
    },
    {
      code: "a { content: ')'; }"
    },
    {
      code: "a { content: '))'; }"
    },
    {
      code: "a { content: '(\t) )'; }"
    },
    {
      code: ":root { --foo: '('; --bar: '))'; }"
    }
  ],

  reject: [
    {
      code: ")",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: " )",
      message: messages.rejected,
      line: 1,
      column: 2
    },
    {
      code: "  )",
      message: messages.rejected,
      line: 1,
      column: 3
    },
    {
      code: "\n)",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "\n\n)",
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: "\r\n)",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "\r\n\r\n)",
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: "\r)",
      message: messages.rejected,
      line: 1,
      column: 2
    },
    {
      code: "\r\r)",
      message: messages.rejected,
      line: 1,
      column: 3
    },
    {
      code: "\t)",
      message: messages.rejected,
      line: 1,
      column: 2
    },
    {
      code: "\t\t)",
      message: messages.rejected,
      line: 1,
      column: 3
    },
    {
      code: "a {)}",
      message: messages.rejected,
      line: 1,
      column: 4
    },
    {
      code: "a { )}",
      message: messages.rejected,
      line: 1,
      column: 5
    },
    {
      code: "a {) }",
      message: messages.rejected,
      line: 1,
      column: 4
    },
    {
      code: "a { ) }",
      message: messages.rejected,
      line: 1,
      column: 5
    },
    {
      code: "a {  )}",
      message: messages.rejected,
      line: 1,
      column: 6
    },
    {
      code: "a {)  }",
      message: messages.rejected,
      line: 1,
      column: 4
    },
    {
      code: "a {  )  }",
      message: messages.rejected,
      line: 1,
      column: 6
    },
    {
      code: "a {   )   }",
      message: messages.rejected,
      line: 1,
      column: 7
    },
    {
      code: "a {\n)}",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a {)\n}",
      message: messages.rejected,
      line: 1,
      column: 4
    },
    {
      code: "a {\n)\n}",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a {\r\n)}",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a {)\r\n}",
      message: messages.rejected,
      line: 1,
      column: 4
    },
    {
      code: "a {\r\n)\r\n}",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a {\t)}",
      message: messages.rejected,
      line: 1,
      column: 5
    },
    {
      code: "a {)\t}",
      message: messages.rejected,
      line: 1,
      column: 4
    },
    {
      code: "a {\t)\t}",
      message: messages.rejected,
      line: 1,
      column: 5
    },
    {
      code: "a { color: var(--prop); })",
      message: messages.rejected,
      line: 1,
      column: 26
    },
    {
      code: "a { color: var(--prop); } )",
      message: messages.rejected,
      line: 1,
      column: 27
    },
    {
      code: "a { color: var(--prop); }\n)",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a { color: var(--prop); }\n\n)",
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: "a { color: var(--prop); }\r\n)",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a { color: var(--prop); }\r\n\r\n)",
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: "a { color: var(--prop); }\r)",
      message: messages.rejected,
      line: 1,
      column: 19
    },
    {
      code: "a { color: var(--prop); }\r\r)",
      message: messages.rejected,
      line: 1,
      column: 28
    },
    {
      code: "a { color: var(--prop); }\t)",
      message: messages.rejected,
      line: 1,
      column: 27
    },
    {
      code: "a { color: var(--prop); }\t\t)",
      message: messages.rejected,
      line: 1,
      column: 28
    },
    {
      code: ")a { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ") a { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")  a { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")\na { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")\n\na { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")\r\na { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")\r\n\r\na { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")\ta { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: ")\t\ta { color: var(--prop); }",
      message: messages.rejected,
      line: 1,
      column: 1
    },
    {
      code: "@media ()) { }",
      message: messages.rejected,
      line: 1,
      column: 10
    },
    {
      code: "@media () ) { a { color: var(--prop); } }",
      message: messages.rejected,
      line: 1,
      column: 11
    },
    {
      code: "@media ()  ) { a { color: var(--prop); } }",
      message: messages.rejected,
      line: 1,
      column: 12
    },
    {
      code: "@media (var(--prop))) { a { color: var(--prop); } }",
      message: messages.rejected,
      line: 1,
      column: 21
    },
    {
      code: "@media (var(--prop) )) { a { color: var(--prop); } }",
      message: messages.rejected,
      line: 1,
      column: 22
    },
    {
      code: "@media ( var(--prop) )) { a { color: var(--prop); } }",
      message: messages.rejected,
      line: 1,
      column: 23
    },
    {
      code: "/* comment */)",
      message: messages.rejected,
      line: 1,
      column: 14
    },
    {
      code: "/* comment */ ) /*comment */",
      message: messages.rejected,
      line: 1,
      column: 15
    },
    {
      code: "/* comment */\n)\n/* comment */",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "/* comment */\r\n)\r\n/* comment */",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "/* comment */\t)\t/* comment */",
      message: messages.rejected,
      line: 1,
      column: 15
    },
    {
      code: "a { color: var(--prop)) }",
      message: messages.rejected,
      line: 1,
      column: 23
    },
    {
      code: "a { color: var(--prop) ) }",
      message: messages.rejected,
      line: 1,
      column: 24
    },
    {
      code: "a { color: var(--prop)  ) }",
      message: messages.rejected,
      line: 1,
      column: 25
    },
    {
      code: "a { color: var(--prop)\n) }",
      message: messages.rejected,
      line: 2,
      column: 1
    },

    {
      code: "a { color: var(--prop)\n\n) }",
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: "a { color: var(--prop)\r\n) }",
      message: messages.rejected,
      line: 2,
      column: 1
    },
    {
      code: "a { color: var(--prop)\r\n\r\n) }",
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: "a { background: url()) }",
      message: messages.rejected,
      line: 1,
      column: 22
    },
    {
      code: "a { background: url( )) }",
      message: messages.rejected,
      line: 1,
      column: 23
    },
    {
      code: "a { background: url( ) ) }",
      message: messages.rejected,
      line: 1,
      column: 24
    },
    {
      code: "a { background: url('')) }",
      message: messages.rejected,
      line: 1,
      column: 24
    },
    {
      code: "a { background: url(' ') ) }",
      message: messages.rejected,
      line: 1,
      column: 26
    },
    {
      code: "a { @media (min-width: var(--prop))) { color: red; } }",
      message: messages.rejected,
      line: 1,
      column: 36
    },
    {
      code: "a { @media (min-width: var(--prop)) ) { color: red; } }",
      message: messages.rejected,
      line: 1,
      column: 37
    },
    {
      code: "a { @media (min-width: var(--prop) )) { color: red; } }",
      message: messages.rejected,
      line: 1,
      column: 37
    },
    {
      code: "a { @media (min-width: var(--prop ))) { color: red; } }",
      message: messages.rejected,
      line: 1,
      column: 37
    },
    {
      code: "a {\n@media (min-width: var(--prop))) { color: red; } }",
      message: messages.rejected,
      line: 2,
      column: 32
    },
    {
      code: "a {\n@media (\nmin-width: var(--prop))) { color: red; } }",
      message: messages.rejected,
      line: 3,
      column: 24
    },
    {
      code: "a {\n@media (\nmin-width: var(--prop)\n)) { color: red; } }",
      message: messages.rejected,
      line: 4,
      column: 2
    },
  ]
});

testRule(rule, {
  ruleName,
  config: [true],
  syntax: "html",

  accept: [
    {
      code: `<div>
<style>
a {
  color: var(--prop);
}
</style>
</div>`
    },
    {
      code: `<div>
<style>
a {
  background: url();
}
</style>
</div>`
    }
  ],

  reject: [
    {
      code: `<div>
<style>
)
</style>
</div>`,
      message: messages.rejected,
      line: 3,
      column: 1
    },
    {
      code: `<div>
<style>
a {
  color: var(--prop))
}
</style>
</div>`,
      message: messages.rejected,
      line: 4,
      column: 21
    }
  ]
});

testRule(rule, {
  ruleName,
  config: [true],
  syntax: "less",

  accept: [
    {
      code: "a { .mixin(); .mixin2; }"
    },
  ],

  reject: [
    {
      code: "a { .mixin());\ncolor: red; }",
      message: messages.rejected,
      line: 1,
      column: 13
    }
  ]
});
```

**Comment by evilebottnawi** [comment]
@alisonailea Sound great idea for rule for me

**Comment by hudochenkov** [comment]
Looks like a good idea! And it meets our [criteria for inclusion](https://stylelint.io/developer-guide/rules/#criteria-for-inclusion). Let's wait for @jeddy3, because he usually sees what other miss :)

Even if it won't make to the core, it would make a good plugin.

**Comment by ntwb** [comment]
Does https://github.com/csstree/stylelint-validator detect this?

Looking at https://csstree.github.io/docs/validator.html many of the examples in the test cases above do throw `Parse error: Identifier is expected` warnings.

**Comment by jeddy3** [comment]
> Does https://github.com/csstree/stylelint-validator detect this?

In its current form, it: 

- will in the value of declarations e.g. functions like `calc`, `url` and `var` etc
- won't in the params of at-rules e.g. `media` and `supports` etc

CSSTree throws a parse error for the former.

It looks like the PostCSS parser itself will throw a [`Unclosed bracket`](https://astexplorer.net/#/gist/a1390ff7bedff5d4438696175f4467fb/90605142c9c2cca76974acd5068e2c556313e852) or [`Unknown word`](https://astexplorer.net/#/gist/a1390ff7bedff5d4438696175f4467fb/6a56ab24c7f0b6a4bba6f2bb2d5ea8631fd4ffbe) parse error for any extra parenthesis outside of the params of an at-rule or the value of a declaration.

In [the VISION](https://github.com/stylelint/stylelint/blob/master/VISION.md#possible-errors) we have:

> Invalid code is best handled by emerging dedicated tools e.g. csstree - a language parser with syntax validation. As a stop-gap, while these tools mature, provide invalid code rules for the simplest of cases.

As the parser itself covers off most instances, I think this will be one of those "simplest cases" and we can add this rule as a stop-gap for the extra parenthesis in the:

1. params of at-rules for everyone
2. value of declarations for users of non-standard syntax

@alisonailea Thank you for offering to contribute this rule. You'll find [guidance](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#creating-a-new-rule) on how to do this in the Developer Guide.

I think you can strip the tests right back, though. Just focus on standard CSS syntax e.g. extra parens in `@media`, `@support`, `calc`, `url` and `var`. You can throw in a couple of SCSS syntax tests for `@include` and `@mixin`, though.

- Name: `no-extra-parenthesis`
- Primary option: `true`
- Secondary options: none
- Message: "Unexpected extra parenthesis"
- Section: "Possible errors" -> "General / Sheet"

**Comment by jeddy3** [comment]
Having thought a little more about this, I think we should limit the scope of this rule to _unbalanced parentheses_. This is the use case @alisonailea originally describes.

The word _extra_ implies _redundant_, which would broaden the scope of this rule to include _balanced but unnecessary parentheses_, e.g. `calc((10px) + (10px))`, and [that isn't viable](https://github.com/stylelint/stylelint/issues/632). This wasn't an issue with semicolons, hence the name `no-extra-semicolons`.

We should also pluralise the rule name to be consistent with `no-extra-semicolons`:

- Name: `no-unbalanced-parentheses`
- Primary option: `true`
- Secondary options: none
- Message: "Unexpected unbalanced parenthesis"
- Section: "Possible errors" -> "General / Sheet"

**Comment by davidmaxwaterman** [comment]
Would this fix the following synthetic case, which just hit me in real code, where I copy/pasted part of a `$0.querySelector(....)` call I did in chrome devtools as a test, but copied a ')' by mistake and didn't notice?
```
$ echo ' ) {
  background: red;
}
' | npx stylelint
$
```

**Comment by jeddy3** [comment]
> Would this fix the following synthetic case

PostCSS [parses this as a selector of `)`](https://astexplorer.net/#/gist/dadf611cbe8e99f41781d5f8a6975eef/245a95d84616d9625be03a9962981bebf189d89a).

We can expand [the scope](https://github.com/stylelint/stylelint/issues/3480#issuecomment-406219672) of the rule to include selectors, so the complete scope becomes:

- selectors of rules
- params of at-rules
- values of declarations

Please consider [contributing](https://stylelint.io/contributing) the rule if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

**Comment by besalraj30** [comment]
Hi

I’d like to work on this  issue. Could you assign it to me?

**Comment by jeddy3** [comment]
@RajvardhanT7747 & @besalraj30, sorry for the delay in replying. We [recently added rules](https://stylelint.io/CHANGELOG#16130---2025-01-12) that overlap with the original proposal so I wanted to wait a while before revisiting this.

Brackets in all constructs except selectors are now checked in other rules. Let's change the scope of this rule to invalid selectors and leverage CSSTree:

```js
import * as csstree from "css-tree";

try {
  csstree.parse("a )", { context: "selector" });
} catch (e) {
  console.error(e);
}
```

```console
SyntaxError: Unexpected input
    at TokenStream.error (file:///Users/jeddy3/Projects/stylelint-test/node_modules/css-tree/lib/parser/create.js:284:19)
    at Module.parse (file:///Users/jeddy3/Projects/stylelint-test/node_modules/css-tree/lib/parser/create.js:340:20)
    at file:///Users/jeddy3/Projects/stylelint-test/scripts/csstree.mjs:4:11
    at ModuleJob.run (node:internal/modules/esm/module_job:271:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:578:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5) {
  source: 'a )',
  offset: 2,
  line: 1,
  column: 3,
  sourceFragment: [Function: sourceFragment],
  formattedMessage: 'Parse error: Unexpected input\n    1 |a )\n---------^'
}
```

Rule details:

- Name: `selector-no-invalid`
- Description: Disallow invalid selectors.
- Message: `Invalid selector "${selector}"`
- Primary option: `true`
- Secondary options: None
- Category: Avoid errors > Invalid
- Included in sharable configs?: Yes, recommended
- Expanded description: This rule considers selector syntax defined within the CSS specifications to be valid.

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

You can use [`at-rule-prelude-no-invalid`](https://github.com/stylelint/stylelint/tree/main/lib/rules/at-rule-prelude-no-invalid) as a blueprint, see https://github.com/stylelint/stylelint/pull/8268.

Let us know if either of you would like to contribute this rule, and I'll label the issue as "wip" for you.
